spring-kafka/src/main/java/org/springframework/kafka/listener/KafkaMessageListenerContainer.java

```diff
@@ -895,6 +899,15 @@ else if (listener instanceof MessageListener) {
 				this.wantsFullRecords = false;
 				this.pollThreadStateProcessor = setUpPollProcessor(false);
 				this.observationEnabled = this.containerProperties.isObservationEnabled();
+
+				if (!AopUtils.isAopProxy(listener)) {
+					BiConsumer<ConsumerRecord<K, V>, RuntimeException> callbackForAsyncFailureQueue =
+							(cRecord, ex) -> {
+								FailedRecordTuple<K, V> failedRecord = new FailedRecordTuple<>(cRecord, ex);
+								this.failedRecords.addLast(failedRecord);
+							};
+					this.listener.setCallbackForAsyncFailure(callbackForAsyncFailureQueue);
```

Comment:
We just worked with @sobychacko about async observation propagation.
And we have figured out that it is really just a `MessagingMessageListenerAdapter` responsibility.
The other `GenericMessageListener` impls are out of that async logic scope and has to deal with an observation on its own, if that does not fit into a normal logic of the listener container.

Perhaps something similar here for those `@RetryableTopic` variants?
I mean no need to worry about `GenericMessageListener` contract and extra hook on it.
Just have a setter on the `MessagingMessageListenerAdapter` and propagate such a callback only in that case.

Reply:
I think so, too. 
But the problem is that multiple type casts are required. 
However, I prefer type casting over introducing unnecessary contracts to the interface, which would cause more significant changes! 😅 

I removed callback interface from both `MessageListener` and `GenericMessageListener` interface!